src/BoundaryConditions/fill_halo_regions_open.jl

```diff
@@ -1,39 +1,50 @@
+struct OrlanskiBoundary{G}
+    tendency::G
+end
+
 #####
 ##### Outer functions for filling halo regions for open boundary conditions.
 #####
 
-# TODO: support true open boundary conditions.
-# For this we need to have separate functions for each of the six boundaries,
-# and need to unroll a loop over the boundary normal direction.
-# The syntax for `getbc` is also different for OpenBoundaryCondition than for others,
-# because the boundary-normal index can vary (and array boundary conditions need to be
-# 3D in general).
+@inline _fill_bottom_halo!(i, j, grid, c, bc::OBC, loc, args...) = @inbounds c[i, j, 1]           = getbc(bc, i, j, grid, args...)
+@inline    _fill_top_halo!(i, j, grid, c, bc::OBC, loc, args...) = @inbounds c[i, j, grid.Nz + 1] = getbc(bc, i, j, grid, args...)
+
+@inline function _fill_south_halo!(i, k, grid, ϕ, bc::BoundaryCondition{<:Open, <:OrlanskiBoundary}, loc, args...)
+    c = max(0, min(1, -bc.condition.tendency[i, 1, k] / (ϕ[i, 2, k] - ϕ[i, 1, k] + eps(0.0))))
 
-@kernel function set_west_or_east_u!(u, i_boundary, bc, grid, args) 
-    j, k = @index(Global, NTuple)
-    @inbounds u[i_boundary, j, k] = getbc(bc, j, k, grid, args...)
+    ϕ[i, 0, k] = (1 - c) * ϕ[i, 0, k] + c * ϕ[i, 1, k]
 end
 
-@kernel function set_south_or_north_v!(v, j_boundary, bc, grid, args)
-    i, k = @index(Global, NTuple)
-    @inbounds v[i, j_boundary, k] = getbc(bc, i, k, grid, args...)
+@inline function _fill_north_halo!(i, k, grid, ϕ, bc::BoundaryCondition{<:Open, <:OrlanskiBoundary}, loc, args...)
+    c = max(0, min(1, -bc.condition.tendency[i, grid.Ny, k] / (ϕ[i, grid.Ny, k] - ϕ[i, grid.Ny - 1, k] + eps(0.0))))
```

Comment:
why not use `@inbounds` here?

also, what's `eps(0.0)`? is this intentional?

Reply:
Probably that should be a parameter of the boundary condition.

We'll have to figure out the logic of the code as well. Perhaps we can add appropriate tendencies within `regularize_boundary_conditions` in the model constructor. Another option is to pass the tendencies into `fill_halo_regions!` which could make sense...

Reply:
I had to have the `eps(0.0)` for when `ϕ[i, grid.Ny, k] - ϕ[i, grid.Ny - 1, k] = 0` for example if you set the field to a constant value